resolwe_bio/processes/expression/featureCounts.py

```diff
@@ -0,0 +1,888 @@
+
+        # for paired-end samples, extend the args with paired-end input options
+        if paired_end and inputs.paired_end.is_paired_end:
+            args.append("-p")
+
+        if paired_end and inputs.paired_end.require_both_ends_mapped:
+            args.append("-B")
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.append("-P")
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.extend(["-d", inputs.paired_end.min_frag_length])
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.extend(["-D", inputs.paired_end.max_frag_length])
+
+        if paired_end and inputs.paired_end.do_not_count_chimeric_fragments:
+            args.append("-C")
+
+        if paired_end and inputs.paired_end.do_not_sort:
+            args.append("--do_not_sort")
+
+        # Run featureCounts
+        return_code, _, _ = Cmd["featureCounts"][args][bam_file] & TEE(retcode=None)
+        if return_code:
+            self.error("Error while running featureCounts.")
+
+        # parse featureCounts output
+        get_gene_counts("featureCounts_rc.txt", "rc.txt", bam_file)
+        get_gene_lenghts("featureCounts_rc.txt", "gene_lengths.txt")
+
+        # Normalize counts
+        rnanorm_args = [
+            "rc.txt",
+            "--gene-lengths",
+            "gene_lengths.txt",
+            "--tpm-output",
+            "tpm.txt",
+            "--cpm-output",
+            "cpm.txt",
+        ]
+        return_code, _, _ = Cmd["rnanorm"][rnanorm_args] & TEE(retcode=None)
+        if return_code:
+            self.error("Error while normalizing counts using rnanorm.")
+
+        # prepare the expression set outputs
+        feature_ids = pd.read_csv(
+            "rc.txt", sep="\t", index_col="FEATURE_ID"
+        ).index.to_list()
+
+        feature_filters = {
+            "source": inputs.annotation.output.source,
+            "species": inputs.aligned_reads.output.species,
+            "feature_id__in": feature_ids,
+        }
+
+        feature_ids_to_names = {
+            f.feature_id: f.name for f in self.feature.filter(**feature_filters)
+        }
+
+        prepare_expression_set(
+            rc="rc.txt",
+            tpm="tpm.txt",
+            cpm="cpm.txt",
+            feature_dict=feature_ids_to_names,
+            outfile_name=f"{name}_expressions",
+        )
+
+        # rename and compress the expression files
+        rename_columns_and_compress("rc.txt", f"{name}_rc.tab.gz")
+        rename_columns_and_compress("tpm.txt", f"{name}_tpm.tab.gz")
+        rename_columns_and_compress("cpm.txt", f"{name}_cpm.tab.gz")
+
+        # Compress the featureCounts output file
+        (
+            Cmd["gzip"]["-c", "featureCounts_rc.txt"]
+            > f"{name}_featureCounts_rc.txt.gz"
+        )()
+
+        # Save the abundance estimates to JSON storage
+        Cmd["expression2storage.py"]("--output", "json.txt", exp_output)
```

Comment:
Move this script into a function.  Eventually this will have to become a beast on its own that will have to reside within `/resolwe_bio/tools/` 

Perhaps something along the lines of

```python
import os
import utils
import json


def expression_to_storage(rc_input, rc_output):
    
    if not (rc_input and os.path.isfile(rc_input)):
        raise ValueError("Specify rc_input")

    def isfloat(value):
        """Check if value is float."""
        try:
            float(value)
            return True
        except ValueError:
            return False

    with utils.gzopen(rc_input) as f:
        # Split lines by tabs
        # Ignore lines without a number in second column
        # Build a dictionary of gene-expression pairs
        exp = {
            "genes": {
                gene_exp[0]: float(gene_exp[1])
                for gene_exp in (l.split("\t") for l in f)
                if len(gene_exp) == 2 and isfloat(gene_exp[1])
            }
        }

    if rc_output:
        with open(file=rc_output, mode="wt") as f:
            json.dump(exp, f)
    
    return rc_output
```

Reply:
Done, with slight simplifications.